File: RentalsDownloader/Rental.cs
namespace RentalsDownloader
{
	using System;
	using System.Collections.Generic;
	using System.IO;
	using System.Linq;
	using System.Text;
	using System.Text.RegularExpressions;
	using System.Xml.Serialization;
	using Common;
	using JetBrains.Annotations;
	using WebRobotFramework;

	[Serializable, PublicAPI]
	public sealed class Rental : DownloadItem
	{
		public static readonly string ReportTemplate;
		static readonly Regex NeighborhoodRegex = new Regex(@"(?:Los Angeles */ *)?([\w\s]+)");
		static readonly Regex NumberRegex = new Regex("(\\d[\\d\\.,]*)");
		static readonly string PhotoTemplate;
		static readonly string RentalTemplate;
		static readonly XmlSerializer XmlSerializer = new XmlSerializer(typeof(Rental));

		static Rental()
		{
			ReportTemplate = File.ReadAllText($@"{RentalsDownloaderService.DownloadPath}template\template.html", Encoding.UTF8);
			RentalTemplate = Regex.Match(ReportTemplate, "<rental>(.*)</rental>", RegexOptions.Singleline).Groups[1].Captures[0].Value;
			PhotoTemplate = Regex.Match(ReportTemplate, "<photo>(.*)</photo>", RegexOptions.Singleline).Groups[1].Captures[0].Value;
		}

		public bool AcceptsCats => Pets != "No pets" && Pets != "Dog ok";
		public string Address { get; set; }
		public List<string> Amenities { get; set; }
		public string Available { get; set; }
		public string Bathrooms { get; set; }
		public string Bedrooms { get; set; }
		public int BedroomsAsInt => (int)GetDoubleValue(Bedrooms);
		public string Called { get; set; }
		public string Comments { get; set; }
		public string Contact { get; set; }
		public string Deposit { get; set; }
		public string Email { get; set; }
		public string Furnished { get; set; }
		public bool HasParking => Parking != "Street parking";
		public override string Id { get; set; }
		public string LeaseType { get; set; }
		public string ListingInfo { get; set; }
		public string ListingType { get; set; }
		public string Location { get; set; }
		public string Neighboorhood => NeighborhoodRegex.Match(Location).Groups[1].Captures[0].Value;
		public string Parking { get; set; }
		public string Pets { get; set; }
		public List<string> Photos { get; set; }
		public string Rating { get; set; } = "";
		public int RatingAsInt => (int)GetDoubleValue(Rating);
		public string Rent { get; set; }
		public double RentAsDouble => GetDoubleValue(Rent);
		public double RentPerSquareFeet => SquareFootageAsInt > 0 && RentAsDouble > 0 ? RentAsDouble / SquareFootageAsInt : double.MaxValue;
		public string SpreadsheetHyperlink => $"HYPERLINK(\"{Url}\",\"Link\")";
		public string SquareFootage { get; set; }
		public int SquareFootageAsInt => (int)GetDoubleValue(SquareFootage);
		public string StructureType { get; set; }
		public string Title { get; set; }
		public string UnitDetails { get; set; }
		public override string Url => $"http://www.westsiderentals.com/listingdetail/{Id}/";
		public string Visited { get; set; }
		public string WalkScore { get; set; }

		public override void Parse(SeleniumInterface selenium)
		{
			Title = GetNodeText(selenium, "//div[contains(@class,'address-info')]//h1");
			ListingInfo = GetNodeText(selenium, "//div[contains(@class,'address-info')]//h3");
			Amenities = selenium.SelectNodes("//div[contains(@class,'address-info')]//li").Select(l => l.InnerText).ToList();
			Photos = selenium.SelectNodes("//div[@id='carousel']//img").Select(l => l.GetAttributeValue("src", "")).ToList();
			Address = GetNodeText(selenium, "//div[contains(@class, 'address-info')]/div[@class='bigText']");
			Contact = GetNodeText(selenium, "//div[contains(@class, 'address-info')]/div[@class='medText']");
			Email = GetNodeText(selenium, "//div[contains(@class, 'address-info')]//a[@class='mailtoLink']");
			WalkScore = GetNodeText(selenium, "//div[@class='walkscore']//span");
			var details = selenium.SelectNodes("//div[@class='col-md-6']").Select(l => Regex.Replace(l.InnerText, "[\r\t\n]", "")).ToList();
			Rent = GetRentalItem(details, "Square Footage");
			Location = GetRentalItem(details, "City/Area");
			Rent = GetRentalItem(details, "Rent");
			SquareFootage = GetRentalItem(details, "Square Footage");
			Bedrooms = GetRentalItem(details, "Bedrooms");
			Pets = GetRentalItem(details, "Pets");
			StructureType = GetRentalItem(details, "Structure Type");
			Parking = GetRentalItem(details, "Parking");
			Available = GetRentalItem(details, "Available");
			Bathrooms = GetRentalItem(details, "Bathrooms");
			Furnished = GetRentalItem(details, "Furnished");
			LeaseType = GetRentalItem(details, "Lease Type");
			UnitDetails = GetRentalItem(details, "Unit Details");
			ListingType = GetRentalItem(details, "Listing Type");
		}

		public void SaveToSpreadsheet(WorkSheet worksheet, int rowIndex)
		{
			worksheet.SetCell(rowIndex, "Address", Address);
			worksheet.SetCell(rowIndex, "Neighborhood", Neighboorhood);
			worksheet.SetCell(rowIndex, "SqFeet", SquareFootageAsInt);
			worksheet.SetCell(rowIndex, "Rooms", BedroomsAsInt);
			worksheet.SetCell(rowIndex, "Price", RentAsDouble);
			worksheet.SetCell(rowIndex, "Structure", UnitDetails);
			worksheet.SetCell(rowIndex, "Comments", Comments);
			worksheet.SetCell(rowIndex, "V", Visited);
			worksheet.SetCell(rowIndex, "C", Called);
			if (Rating != "")
				worksheet.SetCell(rowIndex, "R", RatingAsInt);
			worksheet.SetCellFormula(rowIndex, "Link", $"={SpreadsheetHyperlink}");
			worksheet.SetCellFormula(rowIndex, "Map", $"=HYPERLINK(\"http://maps.google.com/?q={Address}\",\"Map\")");
			worksheet.SetCellFormula(rowIndex, "Redfin", $"=HYPERLINK(\"https://www.google.com/search?q={Address} Redfin\",\"Redfin\")");
			worksheet.SetCell(rowIndex, "Contact", Contact);
			worksheet.SetCell(rowIndex, "D", Amenities.Any(a => a.Contains("dishwasher", StringComparison.OrdinalIgnoreCase)) ? "X" : "");
			worksheet.SetCell(rowIndex, "L", HasLaundry());
		}

		public override void Serialize(StreamWriter writer) => XmlSerializer.Serialize(writer, this);

		public string ToHtml(int index)
		{
			var rentalHtml = RentalTemplate;
			rentalHtml = rentalHtml.Replace("[Url]", Url);
			rentalHtml = rentalHtml.Replace("[Title]", $"{index}. {Title}");
			rentalHtml = rentalHtml.Replace("[Location]", Location);
			rentalHtml = rentalHtml.Replace("[Price]", Rent);
			rentalHtml = rentalHtml.Replace("[SquareFeet]", RentPerSquareFeet < double.MaxValue ? $"{SquareFootage} ({RentPerSquareFeet:F2}$/sq.ft.)" : SquareFootage);
			rentalHtml = rentalHtml.Replace("[Bedrooms]", Bedrooms);
			rentalHtml = rentalHtml.Replace("[Availability]", Available);
			rentalHtml = rentalHtml.Replace("[Pets]", Pets);
			rentalHtml = rentalHtml.Replace("[Type]", StructureType);
			rentalHtml = rentalHtml.Replace("[Parking]", Parking);
			rentalHtml = rentalHtml.Replace("[Address]", Address);
			rentalHtml = rentalHtml.Replace("[Contact]", Contact);
			rentalHtml = rentalHtml.Replace("[Photos]", Photos.Aggregate("", (current, photo) => current + PhotoTemplate.Replace("[PhotoURL]", photo)));
			return rentalHtml;
		}

		static double GetDoubleValue(string value)
		{
			var match = NumberRegex.Match(value);
			if (!match.Success)
				return 0;
			double squareFootage;
			return double.TryParse(match.Groups[1].Captures[0].Value, out squareFootage) ? squareFootage : 0;
		}

		static string GetRentalItem(IEnumerable<string> items, string prefix)
			=> items.FirstOrDefault(attribute => attribute.StartsWith(prefix + ":", StringComparison.Ordinal))?.Substring(prefix.Length + 1).Trim() ?? "";

		string HasLaundry()
		{
			if (!Amenities.Any(a => a.Contains("laundry", StringComparison.OrdinalIgnoreCase) || a.Contains("washer", StringComparison.OrdinalIgnoreCase) || a.Contains("dryer", StringComparison.OrdinalIgnoreCase)))
				return "";
			return Amenities.Any(a => a.Contains("hook", StringComparison.OrdinalIgnoreCase)) ? "H" : "X";
		}
	}
}

File: RentalsDownloader.Service/Rental.fs
module Rental

open Common
open System
open System.IO
open System.Text
open System.Text.RegularExpressions
open System.Xml.Serialization
open WebRobotFramework

let numberRegex = new Regex("(\\d[\\d\\.,]*)")
let neighboorhoodRegex = new Regex(@"(?:Los Angeles */ *)?([\w\s]+)")
let getUrlForId (id : string) = String.Format("http://www.westsiderentals.com/listingdetail/{0}/", id)
let xmlSerializer = new XmlSerializer(typeof<Rental>)

let download (selenium : SeleniumInterface) fileCachePath id = 
    let url = getUrlForId id
    
    let parse() = 
        let getNode xpath = 
            let node = selenium.SelectNode(xpath)
            if node = null then ""
            else node.InnerText.Trim()
        
        let selectNodes xpath f = 
            selenium.SelectNodes(xpath)
            |> Seq.map (f)
            |> Seq.toList
        
        let getDoubleValue s = 
            let m = numberRegex.Match(s)
            if not m.Success then 0.0
            else 
                let parsed, result = Double.TryParse(m.Groups.[1].Captures.[0].Value)
                if parsed then result
                else 0.0
        
        let details = selectNodes "//div[@class='col-md-6']" (fun l -> Regex.Replace(l.InnerText, "[\r\t\n]", ""))
        
        let getRentalItem prefix = 
            let attrib = details |> Seq.tryFind (fun attrib -> attrib.StartsWith(prefix + ":", StringComparison.Ordinal))
            match attrib with
            | Some a -> a.Trim()
            | None -> ""
        
        let location = getRentalItem "City/Area"
        let amenities = selectNodes "//div[contains(@class,'address-info')]//li" (fun l -> l.InnerText)
        let bedrooms = getRentalItem "Bedrooms"
        let parking = getRentalItem "Parking"
        let rent = getRentalItem "Rent"
        let rentDouble = getDoubleValue rent
        let squareFootage = getRentalItem "Square Footage"
        let squareFootageInt = int (getDoubleValue squareFootage)
        let pets = getRentalItem "Pets"
        
        let rentPerSqFt = 
            match (squareFootageInt, rentDouble) with
            | (sqFt, rent) when sqFt > 0 && rent > 0.0 -> Some(rent / double sqFt)
            | _ -> None
        { download = 
              { id = id
                url = url }
          title = getNode "//div[contains(@class,'address-info')]//h1"
          location = location
          neighborhood = neighboorhoodRegex.Match(location).Groups.[1].Captures.[0].Value
          address = getNode "//div[contains(@class, 'address-info')]/div[@class='bigText']"
          listingInfo = getNode "//div[contains(@class,'address-info')]//h3"
          amenities = amenities
          photos = selectNodes "//div[@id='carousel']//img" (fun l -> l.GetAttributeValue("src", ""))
          contact = getNode "//div[contains(@class, 'address-info')]/div[@class='medText']"
          email = getNode "//div[contains(@class, 'address-info')]//a[@class='mailtoLink']"
          walkScore = getNode "//div[@class='walkscore']//span"
          rent = rent
          rentDouble = rentDouble
          squareFootage = squareFootage
          squareFootageInt = squareFootageInt
          rentPerSqFt = rentPerSqFt
          bedrooms = bedrooms
          bedroomsInt = int (getDoubleValue bedrooms)
          pets = pets
          acceptsCats = pets <> "No pets" && pets <> "Dog ok"
          structureType = getRentalItem "Structure Type"
          parking = parking
          hasParking = parking <> "Street parking"
          available = getRentalItem "Available"
          bathrooms = getRentalItem "Bathrooms"
          furnished = getRentalItem "Furnished"
          leaseType = getRentalItem "Lease Type"
          unitDetails = getRentalItem "Unit Details"
          listingType = getRentalItem "Listing Type"
          deposit = "" // todo fill
          called = ""
          comments = ""
          rating = ""
          ratingInt = 0
          visited = "" }
    
    let saveFileToCache rental = 
        use writer = new StreamWriter(String.Format("{0}\\{1}.xml", fileCachePath, rental.download.id))
        xmlSerializer.Serialize(writer, rental)
    
    selenium.GoTo(url) |> ignore
    let rental = parse()
    saveFileToCache rental
    rental

let saveToSpreadsheet (worksheet : WorkSheet) row rental = 
    let strContainsAny (s : string) a = a |> List.exists (fun a' -> s.Contains(a', StringComparison.OrdinalIgnoreCase))
    let setCell a b = worksheet.SetCell(row, a, b)
    let setCellFormula a b = worksheet.SetCellFormula(row, a, b)
    setCell "Address" rental.address
    setCell "Neighborhood" rental.neighborhood
    setCell "SqFeet" rental.squareFootageInt
    setCell "Rooms" rental.bedroomsInt
    setCell "Price" rental.rentDouble
    setCell "Structure" rental.unitDetails
    setCell "Comments" rental.comments
    setCell "V" rental.visited
    setCell "C" rental.called
    if rental.rating <> "" then setCell "R" rental.ratingInt
    setCellFormula "Link" (String.Format("HYPERLINK(\"{0}\",\"Link\")", rental.download.url))
    setCellFormula "Map" (String.Format("=HYPERLINK(\"http://maps.google.com/?q={0}\",\"Map\")", rental.address))
    setCellFormula "Redfin" (String.Format("=HYPERLINK(\"https://www.google.com/search?q={0} Redfin\",\"Redfin\")", rental.address))
    setCell "Contact" rental.contact
    let hasDishwasher = 
        if rental.amenities |> List.exists (fun a -> strContainsAny a [ "dishwasher" ]) then "X"
        else ""
    setCell "D" hasDishwasher
    let hasLaundry = 
        match rental.amenities with
        | a when a |> List.forall (fun a -> not (strContainsAny a [ "laundry"; "washer"; "dryer" ])) -> ""
        | a when a |> List.exists (fun a -> strContainsAny a [ "hook" ]) -> "H"
        | _ -> "X"
    setCell "L" hasLaundry

let reportTemplate = File.ReadAllText(String.Format(@"{0}template\template.html", @"C:\Rentals\", Encoding.UTF8))
let rentalTemplate = Regex.Match(reportTemplate, "<rental>(.*)</rental>", RegexOptions.Singleline).Groups.[1].Captures.[0].Value
let photoTemplate = Regex.Match(reportTemplate, "<photo>(.*)</photo>", RegexOptions.Singleline).Groups.[1].Captures.[0].Value

let toHtml index rental = 
    let strReplace (a : string) b (s : string) = s.Replace(a, b)
    rentalTemplate
    |> strReplace "[Url]" rental.download.url
    |> strReplace "[Title]" (String.Format("{0}. {1}", index, rental.title))
    |> strReplace "[Location]" rental.location
    |> strReplace "[Price]" rental.rent
    |> strReplace "SquareFeet" (match rental.rentPerSqFt with
                                | Some x -> String.Format("{0} ({1:F2}$/sq.ft.)", rental.squareFootage, x)
                                | None -> rental.squareFootage)
    |> strReplace "[Bedrooms]" rental.bedrooms
    |> strReplace "[Availability]" rental.available
    |> strReplace "[Pets]" rental.pets
    |> strReplace "[Type]" rental.structureType
    |> strReplace "[Parking]" rental.parking
    |> strReplace "[Address]" rental.address
    |> strReplace "[Contact]" rental.contact
    |> strReplace "[Photos]" (rental.photos |> List.fold (fun s p -> s + photoTemplate.Replace("[PhotoURL]", p)) "")
